src/crypto/variant4_random_math.h

```diff
@@ -0,0 +1,405 @@
+#ifndef VARIANT4_RANDOM_MATH_H
+#define VARIANT4_RANDOM_MATH_H
+
+// Register size can be configured to either 32 bit (uint32_t) or 64 bit (uint64_t)
+typedef uint32_t v4_reg;
+
+enum V4_Settings
+{
+	// Generate code with minimal theoretical latency = 45 cycles, which is equivalent to 15 multiplications
+	TOTAL_LATENCY = 15 * 3,
+	
+	// Always generate at least 60 instructions
+	NUM_INSTRUCTIONS = 60,
+	
+	// Available ALUs for MUL
+	// Modern CPUs typically have only 1 ALU which can do multiplications
+	ALU_COUNT_MUL = 1,
+
+	// Total available ALUs
+	// Modern CPUs have 4 ALUs, but we use only 3 because random math executes together with other main loop code
+	ALU_COUNT = 3,
+};
+
+enum V4_InstructionList
+{
+	MUL,	// a*b
+	ADD,	// a+b + C, -128 <= C <= 127
+	SUB,	// a-b
+	ROR,	// rotate right "a" by "b & 31" bits
+	ROL,	// rotate left "a" by "b & 31" bits
+	XOR,	// a^b
+	RET,	// finish execution
+	V4_INSTRUCTION_COUNT = RET,
+};
+
+// V4_InstructionCompact is used to generate code from random data
+// Every random sequence of bytes is a valid code
+//
+// There are 8 registers in total:
+// - 4 variable registers
+// - 4 constant registers initialized from loop variables
+//
+// This is why dst_index is 2 bits
+struct V4_InstructionCompact
```

Comment:
AFAIK bit fields are generally not very portable. There should be at least a `static_assert` to make sure `sizeof(V4_InstructionCompact) == 1`. But it's safer to implement it manually with shift/and.

Reply:
Yes, it's implementation-defined in C, but I haven't seen modern compilers that don't pack it in a single byte. I'll rewrite it anyway. Can't use `static_assert` because it's available only in C++.

Reply:
You can do "static assert" in C like this:

```c
enum {
	assert1 = 1 / (sizeof(V4_InstructionCompact) == 1)
};
```

Reply:
If willing to assume C11 -> https://en.cppreference.com/w/c/language/_Static_assert